chapters/part3/fairness.md

```diff
@@ -0,0 +1,733 @@
+        # don't use dump_metrics() because want to log regardless
+        with open(args.baseline_output_file, "w") as fd:
+            fd.write(metrics_json)
+    logger.info("Metrics: %s", metrics_json)
+
+    if create_diff:
+        diff = diff_tool.get_diff()
+        diff_json = json.dumps(diff, indent=2)
+        if args.predictions_diff_output_file:
+            with open(args.predictions_diff_output_file, "w") as fd:
+                fd.write(diff_json)
+        logger.info("Predictions diff: %s", diff_json)
+
+    logger.info("Finished evaluating.")
+
+    return bias_mitigated_metrics, baseline_metrics
+```
+
+This subcommand can be run using: `allennlp evaluate-bias-mitigation /tmp/snli/model.tar.gz https://storage.googleapis.com/allennlp-public-models/snli-roberta.2021-03-11.tar.gz https://storage.googleapis.com/allennlp-public-models/binary-gender-bias-mitigated-snli-dataset.jsonl 32 --bias-mitigated-output-file bm.txt --baseline-output-file bl.txt --predictions-diff-output-file diff.txt`.
+
+4. Inspecting `bm.txt` and `bl.txt` will reveal the following `NaturalLanguageInference` scores for the bias-mitigated, finetuned model and baseline, pretrained model:
+
+```python
+"bias_mitigated_model_metrics": {
+  "net_neutral": 0.6417539715766907,
+  "fraction_neutral": 0.7002295255661011,
+  "threshold_0.5": 0.6902161836624146,
+  "threshold_0.7": 0.49243637919425964
+}, 
+"baseline_model_metrics": {
+  "net_neutral": 0.49562665820121765,
+  "fraction_neutral": 0.5068705677986145,
+  "threshold_0.5": 0.47600528597831726,
+  "threshold_0.7": 0.3036800026893616
+}
+```
+The `NaturalLanguageInference` scores for the bias-mitigated, finetuned model are much closer to 1 than they are for the baseline, pretrained model, which suggests lower binary gender bias with respect to occupations, as bias will result in a higher probability of entailment or contradiction. `diff.txt` allows us to see specific examples for which the bias-mitigated, finetuned model chose neutral but the baseline, pretrained model didn't:
+
+```python
+{
+    "sentence_pair": "<s>An accountant can afford a computer.</s></s>A gentleman can afford a computer.</s><pad><pad><pad><pad>",
+    "bias_mitigated_label": "neutral",
+    "baseline_label": "entailment"
+}
+```
+
+We have made the [bias-mitigated, finetuned large RoBERTA model for SNLI](https://storage.googleapis.com/allennlp-public-models/binary-gender-bias-mitigated-snli-roberta.2021-05-20.tar.gz) publicly available with a [model card](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/modelcards/pair-classification-binary-gender-bias-mitigated-roberta-snli.json).
+
+</exercise>
+
+<exercise id="6" title="Next Steps">
+
+**Friendly reminder:** While the fairness and bias mitigation tools provided by `allennlp.fairness` help reduce and control biases and increase how equitably models perform, they do <em>not</em> by any means remove biases or make models entirely fair.
```

Comment:
How should users react and follow up on this disclaimer?  I would reframe "by any means remove biases" to "completely remove bias".